dlgr/griduniverse/maze.py

```diff
@@ -0,0 +1,119 @@
+import itertools
+import random
+
+
+def generate(rows, columns):
+
+    c = (columns - 1) / 2
+    r = (rows - 1) / 2
+
+    visited = [[0] * c + [1] for _ in range(r)] + [[1] * (c + 1)]
+    ver = [["* "] * c + ['*'] for _ in range(r)] + [[]]
+    hor = [["**"] * c + ['*'] for _ in range(r + 1)]
+
+    sx = random.randrange(c)
+    sy = random.randrange(r)
+    visited[sy][sx] = 1
+    stack = [(sx, sy)]
+    while len(stack) > 0:
+        (x, y) = stack.pop()
+        d = [
+            (x - 1, y),
+            (x, y + 1),
+            (x + 1, y),
+            (x, y - 1)
+        ]
+        random.shuffle(d)
+        for (xx, yy) in d:
+            if visited[yy][xx]:
+                continue
+            if xx == x:
+                hor[max(y, yy)][x] = "* "
+            if yy == y:
+                ver[y][max(x, xx)] = "  "
+            stack.append((xx, yy))
+            visited[yy][xx] = 1
+
+    # Convert the maze to a list of wall cell positions.
+    the_rows = ([j for i in zip(hor, ver) for j in i])
+    the_rows = [list("".join(j)) for j in the_rows]
+    maze = [item == '*' for sublist in the_rows for item in sublist]
+    positions = []
+    for idx, value in enumerate(maze):
+        if value:
+            positions.append([idx / columns, idx % columns])
+
+    return positions
+
+
+def prune(walls, density, contiguity):
+    """Prune walls to a labyrinth with the given density and contiguity."""
+    num_to_prune = int(round(len(walls) * (1 - density)))
+    num_pruned = 0
+    count = 0
+    while num_pruned < num_to_prune:
+        count += 1
+        to_prune = _classify_terminals(
+            walls, limit=num_to_prune - num_pruned
+        )
+        walls = [w for i, w in enumerate(walls) if i not in to_prune]
+        if len(to_prune) == 0:
+            break
+        num_pruned += len(to_prune)
+
+    num_to_prune = int(round(len(walls) * (1 - contiguity)))
+    to_prune = set(random.sample(range(len(walls)), num_to_prune))
+    walls = [w for i, w in enumerate(walls) if i not in to_prune]
+
+    return walls
+
+
+def _classify_terminals(walls, limit=None):
+    found = [set()]
+    unmatched = {}
+    enumerated_walls = tuple(enumerate(walls))
+    position_map = {tuple(w.position): i for i, w in enumerated_walls}
+
+    for i, w in enumerated_walls:
+        neighbors = set()
+        for adj in ([1, 0], [-1, 0], [0, 1], [0, -1]):
+            neighbor = tuple(p1 + p2 for p1, p2 in zip(w.position, adj))
```

Comment:
It's a little confusing that you need to use the `Wall` class from experiment.py to call the prune function, should that be part of this module too?

Reply:
That gave me an immediate icky feeling also, but I'm more inclined to make `prune()` use the same position lists that `generate()` creates, and then convert the final result to `Wall`s back in `experiment.labyrinth()`. My hope was to ultimately have some more comprehensible (and testable) intermediate steps, and a very simple, non-experiment-specific representation of positions seems right for this (I was thinking `namedtuple`s would be a nice alternative to the [0, 1] lists, but I think we're performing arithmetic on them, so that looked like more trouble that than it was worth.) What do you think about that?

Reply:
The `Wall` class is pretty trivial, why not just have it live in `maze.py`. Walls in GU are essentially only meaningful in the context of a maze anyway (you can enable wall building in worlds without mazes, but it's kind of an edge case I think).